src/main/java/net/minestom/server/instance/block/predicate/DataComponentPredicate.java

```diff
@@ -0,0 +1,376 @@
+        }
+    }
+
+    record StoredEnchantment(
+            @Nullable List<DynamicRegistry.Key<net.minestom.server.item.enchant.Enchantment>> enchantments,
+            Range.Int levels) {
+
+        public static Codec<StoredEnchantment> CODEC = StructCodec.struct(
+                "enchantments", net.minestom.server.item.enchant.Enchantment.CODEC.listOrSingle().optional(), StoredEnchantment::enchantments,
+                "levels", DataComponentPredicates.INT_RANGE_CODEC.optional(), StoredEnchantment::levels,
+                StoredEnchantment::new
+        );
+
+        public boolean test(DataComponent.Holder holder) {
+            EnchantmentList holderEnchants = holder.get(DataComponents.STORED_ENCHANTMENTS);
+            if (holderEnchants == null) {
+                return false;
+            }
+            if (enchantments == null) {
+                // If `enchantments` is not specified, the predicate returns true when any enchantment matches the specified `level`
+                return holderEnchants.enchantments().entrySet().stream().anyMatch(entry -> levels == null || levels.inRange(entry.getValue()));
+            }
+            for (DynamicRegistry.Key<net.minestom.server.item.enchant.Enchantment> e : enchantments) {
+                if (holderEnchants.has(e) && (levels == null || levels.inRange(holderEnchants.level(e)))) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+    }
+
+    record Potions(List<PotionType> potionTypes) implements DataComponentPredicate {
+        public static Codec<Potions> CODEC = PotionType.CODEC.listOrSingle().transform(Potions::new, Potions::potionTypes);
+
+        @Override
+        public boolean test(DataComponent.Holder holder) {
+            var potion = holder.get(DataComponents.POTION_CONTENTS);
+            if (potion == null) return false;
+            return potionTypes.contains(potion.potion());
+        }
+    }
+
+    record CustomData(BinaryTag nbt) implements DataComponentPredicate {
+        public static Codec<CustomData> CODEC = Codec.NBT.transform(CustomData::new, CustomData::nbt);
+
+        @Override
+        public boolean test(DataComponent.Holder holder) {
+            net.minestom.server.item.component.CustomData other = holder.get(DataComponents.CUSTOM_DATA);
+            return other != null && other.nbt().equals(nbt);
+        }
+    }
+
+    record Container(CollectionPredicate<ItemStack, ItemPredicate> items) implements DataComponentPredicate {
+        public static Codec<Container> CODEC = StructCodec.struct(
+                "items", CollectionPredicate.createCodec(ItemPredicate.CODEC), Container::items,
+                Container::new
+        );
+
+        @Override
+        public boolean test(DataComponent.Holder holder) {
+            Iterable<ItemStack> itemStacks = holder.get(DataComponents.CONTAINER);
+            return items().test(itemStacks);
```

Comment:
This needs to exclude air (or otherwise air needs to always pass/not affect the results).